mediatek: filogic: add Teltonika RUTC50 support

simsasss opened the pull request:
This pull request add support for Teltonika RUTC50.
The Teltonika RUTC50 is an industrial mobile router, it supports WiFi 6, have 5G REL 16 modem and external USB port.

## Specification:
- MediaTek MT7981A SoC
- 512 MB RAM
- 16MB SPI NOR Flash
- 512MB NAND (split in half for firmware fallback)
- 5x 10/100/1000 Mbps Ethernet, with passive PoE support on LAN1
- WLAN      : MediaTek dual-band WiFi 6
  - 2.4 GHz : b/g/n/ax, MIMO 2x2
  - 5 GHz   : a/n/ac/ax, MIMO 2x2
- Quectel RG520N-EB 5G R16 modem
- 2.0 USB Type-A HOST port
- 1x Digital input
- 1x Digital output
- 2x SIM slot (can be swapped via AT commands)

GPIO:
- 1 button (Reset)
- 13 LEDs (power, 4x WAN status, Wifi 2G, Wifi 5G, 3G, 4G, 5G, RSSI 1,2,3)
- 3 Modem control (power button, reset, status)
- 1 USB reset
- 1 Digital input
- 1 Digital output

## Installation
Flashing via OEM WebUI:
1. Download the firmware image *-squashfs-factory.bin
2. Upload firmware image via OEM WebUI firmware update, do not keep settings

To revert back to OEM firmware:
https://wiki.teltonika-networks.com/view/Bootloader_menu

## Additional info:
Use "ModemManager" to establish mobile data connection.

simsasss committed:
mediatek: filogic: add Teltonika RUTC50 support

Specification:
- MediaTek MT7981A SoC
- 512 MB RAM
- 16MB SPI NOR Flash
- 512MB NAND (split in half for firmware fallback)
- 5x 10/100/1000 Mbps Ethernet, with passive PoE support on LAN1
- WLAN      : MediaTek dual-band WiFi 6
  - 2.4 GHz : b/g/n/ax, MIMO 2x2
  - 5 GHz   : a/n/ac/ax, MIMO 2x2
- Quectel RG520N-EB 5G R16 modem
- 2.0 USB Type-A HOST port
- 1x Digital input
- 1x Digital output
- 2x SIM slot (can be swapped via AT commands)

GPIO:
- 1 button (Reset)
- 13 LEDs (power, 4x WAN status, Wifi 2G, Wifi 5G, 3G, 4G, 5G, RSSI 1,2,3)
- 3 Modem control (power button, reset, status)
- 1 USB reset
- 1 Digital input
- 1 Digital output

Flashing via OEM WebUI:
1. Download the firmware image *-squashfs-factory.bin
2. Upload firmware image via OEM WebUI firmware update, do not keep settings

To revert back to OEM firmware:
https://wiki.teltonika-networks.com/view/Bootloader_menu

Use "ModemManager" to establish mobile data connection.

Signed-off-by: Simonas Tamošaitis <simsasss@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/19401
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

Djfe commented:
do you know why the device has 3 external wifi antennas? (odd, but even number of mimo 2x2)

Djfe commented:
build mediatek/filogic/teltonika_rutc50

hauke commented on target/linux/mediatek/filogic/base-files/etc/board.d/01_leds:
```diff
@@ -187,6 +187,10 @@ smartrg,sdg-8733a)
 	ucidef_set_led_netdev "wan-orange" "WAN" "mdio-bus:08:orange:wan" "wan" "link_100 link_1000"
 	ucidef_set_led_netdev "wan-white" "WAN" "mdio-bus:08:white:wan" "wan" "link_10000"
 	;;
+teltonika,rutc50)
+	ucidef_set_led_netdev "wifi-2g" "wifi-2.4G" "green:wlan-2ghz" "phy0-ap0" "link tx rx"
+	ucidef_set_led_netdev "wifi-5g" "wifi-5G" "green:wlan-5ghz" "phy1-ap0" "link tx rx"
+	;;
```
This should not be needed with the `linux,default-trigger = "phy0tpt";` entry in the device tree.

hauke commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -96,6 +96,20 @@ define Build/cetron-header
 	rm $@.tmp
 endef
 
+define Build/append-teltonika-metadata
+	echo \
+			'{ \
+					"metadata_version": "1.1", \
+					"compat_version": "1.0", \
+					"version": "OpenWrt", \
+					"device_code": [".*"], \
+					"hwver": [".*"], \
+					"batch": [".*"], \
+					"serial": [".*"], \
+					"supported_devices":["teltonika,rutc"] \
+			}' | fwtool -I - $@
+endef
+
```
There is similar functions in `target/linux/ipq40xx/image/generic.mk` could you please create a common one in `include/image-commands.mk` and make use of that in both targets. the ath79 target also has some code which looks similar, see `target/linux/ath79/image/generic.mk` .

hauke commented:
As far as I understood there is one version with SPI NAND flash and one version with eMMC. Do you plan to add support for the eMMC version too?

hauke commented on target/linux/mediatek/dts/mt7981a-teltonika-rutc50.dts:
```diff
@@ -0,0 +1,483 @@
+		mediatek,nmbm;
+		mediatek,bmt-max-ratio = <1>;
+		mediatek,bmt-max-reserved-blocks = <64>;
+
```
Some other devices with this chip are using `spi-cal-enable` is this not needed for your board?

simsasss replied:
Calibration fails, so I think its not.
[    0.987620] spi-nand spi1.0: calibration failed, fallback to default: 0x2

simsasss committed:
images: move append-teltonika-metadata to image-commands.mk

Move append-teltonika-metadata to image-commands.mk and unify over different targets.
This method can be used to create valid "factory" images for most of Teltonika devices.

Signed-off-by: Simonas Tamošaitis <simsasss@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/19401
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

simsasss commented on target/linux/mediatek/dts/mt7981a-teltonika-rutc50.dts:
```diff
@@ -0,0 +1,483 @@
+
+		conf-pu {
+			pins = "SPI0_CS", "SPI0_HOLD", "SPI0_WP";
+			drive-strength = <MTK_DRIVE_8mA>;
```
I don't know how and don't have tools to measure signal quality, but I changed to 4mA, device working fine for few weeks.

hauke commented:
The CI has some problems:
The mac os tools build was fixed in 2876106eb43d9eb04533eec2f2bfbb4574fa2b5d.
Please add a description to all commits.

hauke commented on target/linux/mediatek/dts/mt7981a-teltonika-rutc50.dts:
```diff
@@ -0,0 +1,534 @@
+		stdout-path = "serial0:115200n8";
+	};
+
+	memory {
+		reg = <0 0x40000000 0 0x10000000>;
+	};
```
Please change to:
```
	memory@40000000 {
		reg = <0 0x40000000 0 0x10000000>;
		device_type = "memory";
	};
```

simsasss commented:
> does anyone know whether "order code" equals: there are two revisions of this device? <img alt="grafik" width="764" height="205" src="https://private-user-images.githubusercontent.com/5685512/483871283-b9be567b-8919-4ac5-982f-05482ed82131.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTY1NTE3NjIsIm5iZiI6MTc1NjU1MTQ2MiwicGF0aCI6Ii81Njg1NTEyLzQ4Mzg3MTI4My1iOWJlNTY3Yi04OTE5LTRhYzUtOTgyZi0wNTQ4MmVkODIxMzEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDgzMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA4MzBUMTA1NzQyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MzY0ZTYzZDljYzVmNGM4NjhhYTQxYjY5ZTg4ZTk5NDc4ZGRlMDljYmUyNDE3Y2RiYjUwN2FjZTlkMmE1ZmZjYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.mvevOBaqjw5URGI3oCZK38FBOXmu8jc3BavQPQnS5bw"> source: https://teltonika-networks.com/de/products/routers/rutc50

Yes, emmc version released recently, it will need another dts.
